fedbiomed/transport/node_agent.py

```diff
@@ -74,31 +84,64 @@ def id(self) -> str:
         """
         return self._id
 
-    async def set_active(self) -> None:
-        """Updates node status as active"""
+    async def flush(self, request_id: str, stopped: bool = False) -> None:
+        """Flushes processed reply
 
-        async with self._status_lock:
+        Args:
+            request_id: request ID for which the replies should be flushed
+            stopped: the request was stopped during processing
+        """
 
-            # Inform user that node is online again
-            if self._status == NodeActiveStatus.DISCONNECTED:
-                logger.info(f"Node {self._id} is back online!")
+        async with self._replies_lock:
```

Comment:
Hi @srcansiz 
I don't really understand what the `self._stopped_request_ids` is for ?
(nevertheless, this does not introduce bug - so it is at worst useless - so it's minor issue).

1. it is only used to do a distinct `logger.warning()` in `on_reply` - useful ?
2. this `logger.warning()` does not seem to me very significant because the `self._stopped_request_ids` contains both (1) ID of the request if this node answered before a node stopped the request (2) ID of the request if this node didn't answered before a node stopped the request

IMO the `logger.warning()` would be different in case (1) and (2), which would require something like:
```
        async with self._replies_lock:
           if stopped and self._replies[request_id]['reply'] is None:
               async with self._stopped_request_ids_lock:
                    self._stopped_request_ids.append(request_id)

            self._replies.pop(request_id, None)
```

So finally for me, though it would gain to:
a. either remove completely `self._stopped_request_ids` and associated lock
b. or improve it (eg see above)

Reply:
Hi Marc, 

I think naming was confusing for `self._stopped_request_ids`. It only saves the request ids that has not answered before stopping request. Please look `if stopped and self._replies[request_id]['reply'] is None:`. For example, let's assume that there is strategy for StopOnDisconnect

- Node 1 -> Starts processing request
- Node 2 -> Starts processing request
- Node 3 -> Starts processing request
Then, 

- Node 3 -> Disconnects 
- Federated request stops  

There Node 2 and Node 1 still process the request, however the reply callback is flushed due to stopped request. Then, Node 2 and Node 1 sends their replies. In that case, researcher will assume that a reply received with unexpected request ID.  However, it was a reply for federated request that is stopped.   `self._stopped_request_ids` keeps this in memory and print proper message for those situations.  

If you think that is not necessary I am okay with removing it  but I don't see any improvement.  The improvement should be on the node side where the jobs for canceled federated requests should be canceled on the node side as well.

Reply:
Pushed the proposed modification to handle the case of duplicate replies for stopped requests.